### PR TITLE
https: soWatcher, shared_libraries use a pathIdentifier as key of ELF binaries

### PR DESCRIPTION
**Pull request by nplanel**
### What does this PR do?

* pathIdentifier is an unique id of and ELF (system-wide) as it contain dev and inode as key.
* New Unregister path, thanks to processMonitor that recieve process EXIT event and unregister the uprobe (maintained by a refcount)
* ebpf UID use pathIdentifier as source of truth and use wider alphabet (base64), specially because the UID is limited (5 chars)

### Motivation

Follow up on [#incident-16860](https://dd.slack.com/archives/C044LV5NP1T), [#incident-18347 ](https://dd.slack.com/archives/C04JV29KQUS)

### Describe how to test/QA your changes

```
system_probe_config:
  log_level: debug
  enable_runtime_compiler: true
network_config:
  enable_http_monitoring: true
  enable_https_monitoring: true
```

1/ run system-probe on Ubuntu 22.04 
 * In a container, run wget and or curl but keep them stopped (Ctrl-Z) (to prefetch the libssl.so/crypto.so/gnutls.so loaded)
 * `wget https://httpbin.org/anything/foo/200`
 * `sudo curl -s --unix-socket /opt/datadog-agent/run/sysprobe.sock http://unix/network_tracer/connections | tee cnx | jq '.conns[].tags , .tags'`
You should see openssl tags : `tls.library:openssl`
 * `sudo curl -s --unix-socket /opt/datadog-agent/run/sysprobe.sock http://unix/network_tracer/debug/http_monitoring`

2/ on staging or other cluster
 * create an image (you can reuse the same image)
 * enable https_monitoring like in [PR](https://github.com/DataDog/k8s-datadog-agent-ops/pull/1572/files)
 * deploy in in staging (oddish-a,b,c)
 * look at USM correctness dashboard
   *  you shall see more https traffic `ingress-haproxy-web`
 * restart some ingress-haproxy pods (ask on #fabric)
    * look if any pods are stuck in Restarting state
    * could be helpful
       *  `watch -n 10 'kubectl --context=oddish-a.us1.staging.dog get pod -o wide -A | grep Terminating'`
       * [incident dashboard](https://ddstaging.datadoghq.com/notebook/4121966/-incident-17876?cell_id=lhiycj1d&range=172800000&tpl_var_datacenter=us1.staging.dog&tpl_var_kube_cluster_name=oddish-%2A&from_ts=1670343653539&start=1672930020000&to_ts=1670516453539&live=false)

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Comment by nplanel** [comment]
moving to 7.43 as the fix will be more complex

**Comment by guyarb** [comment]
After the 2 incident with had with that solution, I think we should describe a QA steps & remove `skip-qa` tag

File: pkg/network/protocols/http/shared_libraries.go [hunk 1]
```diff
@@ -35,91 +37,189 @@ func (l *libPath) Bytes() []byte {
...
 
-type seenKey struct {
-	pid, path string
+type soRegistry struct {
+	m     sync.Mutex
+	byID  map[pathIdentifier]*soRegistration
+	byPID map[uint32]*soRegistration
+
+	// if we can't register a uprobe we don't try more than once
+	blacklistByID map[pathIdentifier]struct{}
```
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries.go:102]:
```suggestion
	blockListByID map[pathIdentifier]struct{}
```

File: pkg/network/protocols/http/shared_libraries.go [hunk 2]
```diff
@@ -35,91 +37,189 @@ func (l *libPath) Bytes() []byte {
...
+			byID:          make(map[pathIdentifier]*soRegistration),
+			byPID:         make(map[uint32]*soRegistration),
+			blacklistByID: make(map[pathIdentifier]struct{}),
 		},
 	}
 }
 
+type soRegistration struct {
+	pathID pathIdentifier
+	//mTime  uint64 TODO(nplanel) reload uprobes if the file changed (dnf upgrade)
```
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries.go:128]: is it still relevant?

File: pkg/network/protocols/http/shared_libraries.go [hunk 3]
```diff
@@ -35,91 +37,189 @@ func (l *libPath) Bytes() []byte {
...
+				}
+
+				w.registry.Register(root, m.Path, uint32(pid), r)
+			}
+		}
+
+		return nil
+	})
+
+	err := w.processMonitor.Initialize()
```
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries.go:195]: calling `WithAllProcs` (sync) and only then `w.processMonitor.Initialize()` means that we can loose processes spawned during that time

**Review by guyarb — CHANGES_REQUESTED** [review]

File: pkg/network/protocols/http/ebpf_ssl.go [hunk 4]
```diff
@@ -362,9 +359,9 @@ func (o *sslProgram) Stop() {
 	o.perfHandler.Stop()
 }
 
-func addHooks(m *errtelemetry.Manager, probes []manager.ProbesSelector) func(string) error {
-	return func(libPath string) error {
-		uid := getUID(libPath)
+func addHooks(m *errtelemetry.Manager, probes []manager.ProbesSelector) func(pathIdentifier, string, string) error {
+	return func(id pathIdentifier, root string, path string) error {
```
[the remark this answers is not in the document]
Comment (nplanel) [reply pkg/network/protocols/http/ebpf_ssl.go:363]: it will be associated with path, like `BinaryPath: root+path` was writtening a

File: pkg/network/protocols/http/shared_libraries.go [hunk 5]
```diff
@@ -35,91 +37,189 @@ func (l *libPath) Bytes() []byte {
...
+	return fmt.Sprintf("dev/inode %d.%d/%d", unix.Major(p.dev), unix.Minor(p.dev), p.inode)
+}
+
+func (p *pathIdentifier) Key() string {
+	b := make([]byte, 16)
+	binary.LittleEndian.PutUint64(b, p.dev)
+	binary.LittleEndian.PutUint64(b[8:], p.inode)
+	m := murmur3.Sum64(b)
+	binary.LittleEndian.PutUint64(b, m)
+	return base64.StdEncoding.EncodeToString(b[:8])
```
[the remark this answers is not in the document]
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:56]: b is just local scratch pad here

File: pkg/network/protocols/http/shared_libraries.go (hunk 2 above)
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:128]: yes we don't support upgrade
I created a jira ticket https://datadoghq.atlassian.net/browse/USMO-172?atlOrigin=eyJpIjoiZGUzMGUyY2MzMWRkNDFjZGE3YjBhZWJkM2YyMDU3YzciLCJwIjoiaiJ9

File: pkg/network/tracer/tracer_usm_linux_test.go [hunk 6]
```diff
@@ -159,6 +159,7 @@ func TestHTTPSViaLibraryIntegration(t *testing.T) {
 	if !httpsSupported(t) {
 		t.Skip("HTTPS feature not available/supported for this setup")
 	}
+	t.Skip("we don't support fast process start/stop yet as /proc/pid/root pass likely doesn't exist when we try to add hook")
```
Comment (nplanel) [inline pkg/network/tracer/tracer_usm_linux_test.go:162]: This issue here is that we scan to slowy the ELF, same issue than goTLS

File: pkg/network/protocols/http/shared_libraries.go [hunk 7]
```diff
@@ -35,118 +37,213 @@ func (l *libPath) Bytes() []byte {
...
+
+func (w *soWatcher) processExit(pid uint32) {
+	w.registry.Unregister(pid)
+}
+
 // Start consuming shared-library events
 func (w *soWatcher) Start() {
-	seen := make(map[seenKey]struct{})
-	w.sync()
+	thisPID, _ := util.GetRootNSPID()
```
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries.go:162]: why do you ignore the error?

File: pkg/network/protocols/http/shared_libraries.go [hunk 8]
```diff
@@ -35,118 +37,213 @@ func (l *libPath) Bytes() []byte {
...
+		}
+
+		root := fmt.Sprintf("%s/%d/root", w.procRoot, pid)
+		for _, m := range *mmaps {
+			for _, r := range w.rules {
+				if !r.re.MatchString(m.Path) {
+					continue
+				}
+
+				w.registry.Register(root, m.Path, uint32(pid), r)
```
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries.go:189]: if we matched a rule, is there a reason to keep checking the other rules?

File: pkg/network/protocols/http/shared_libraries_test.go [hunk 9]
```diff
@@ -64,11 +71,66 @@ func TestSharedLibraryDetection(t *testing.T) {
 	assert.Equal(t, fpath, pathDetected)
 }
 
+func TestSharedLibraryDetectionWithRoot(t *testing.T) {
+	t.Skip("skip for the moment as some distro are not friendly with busybox package")
+
+	tdir := t.TempDir()
```
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries_test.go:77]: Use indicative names so other people can easily read the code
```suggestion
	tempDir := t.TempDir()
```

File: pkg/network/protocols/http/shared_libraries_test.go [hunk 10]
```diff
@@ -64,11 +71,66 @@ func TestSharedLibraryDetection(t *testing.T) {
...
+	watcher := newSOWatcher(perfHandler,
+		soRule{
+			re:         regexp.MustCompile(`fooroot.so`),
+			registerCB: callback,
+		},
+	)
+	watcher.Start()
+
+	time.Sleep(10 * time.Millisecond)
+	//	simulateOpenAt(fpath)
```
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries_test.go:114]: it the code is not relevant, remove it
if it is relevant, uncomment

File: pkg/network/protocols/http/shared_libraries_test.go [hunk 11]
```diff
@@ -64,11 +71,66 @@ func TestSharedLibraryDetection(t *testing.T) {
...
+	o, err := exec.Command("unshare", "--fork", "--pid", "-R", root, "/ash", "-c", fmt.Sprintf("sleep 1 > %s", libpath)).CombinedOutput()
+	t.Log(string(o))
```
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries_test.go:116]: is it useful for the tests?

File: pkg/network/protocols/http/shared_libraries_test.go [hunk 12]
```diff
@@ -64,11 +71,66 @@ func TestSharedLibraryDetection(t *testing.T) {
...
+	assert.NoError(t, err)
```
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries_test.go:117]: you are mixing between `require.NoError` and `assert.NoError`
please stick with one of them

File: pkg/network/protocols/http/shared_libraries_test.go [hunk 13]
```diff
@@ -64,11 +71,66 @@ func TestSharedLibraryDetection(t *testing.T) {
...
+	root := filepath.Join(tdir, "root")
+	err := os.MkdirAll(root, 0755)
+	require.NoError(t, err)
+
+	libpath := "/fooroot.so"
+
+	simulateOpenAt(root + libpath)
+	err = exec.Command("cp", "/usr/bin/busybox", root+"/ash").Run()
+	require.NoError(t, err)
+	err = exec.Command("cp", "/usr/bin/busybox", root+"/sleep").Run()
```
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries_test.go:87]: that's not really "sleep", the name is super confusing and does not give the ability to understand the flow of the test

File: pkg/network/protocols/http/ebpf_ssl.go [hunk 14]
```diff
@@ -388,7 +385,7 @@ func addHooks(m *errtelemetry.Manager, probes []manager.ProbesSelector) func(str
 
 				newProbe := &manager.Probe{
 					ProbeIdentificationPair: identifier,
-					BinaryPath:              libPath,
+					BinaryPath:              root + path,
```
Comment (guyarb) [inline pkg/network/protocols/http/ebpf_ssl.go:388]:
```suggestion
					BinaryPath:              filepath.Join(root, path),
```

File: pkg/network/protocols/http/shared_libraries.go [hunk 15]
```diff
@@ -35,118 +37,213 @@ func (l *libPath) Bytes() []byte {
...
+}
+
+func (p *pathIdentifier) Key() string {
+	b := make([]byte, 16)
+	binary.LittleEndian.PutUint64(b, p.dev)
+	binary.LittleEndian.PutUint64(b[8:], p.inode)
+	m := murmur3.Sum64(b)
+	bsum := make([]byte, 8)
+	binary.LittleEndian.PutUint64(bsum, m)
+	return base64.StdEncoding.EncodeToString(bsum)
```
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries.go:57]: please use meaningful names for the variables

File: pkg/network/protocols/http/ebpf_ssl.go [hunk 16]
```diff
@@ -431,14 +428,8 @@ func removeHooks(m *errtelemetry.Manager, probes []manager.ProbesSelector) func(
...
-func getUID(libPath string) string {
-	sum := murmur3.StringSum64(libPath)
-	hash := strconv.FormatInt(int64(sum), 16)
-	if len(hash) >= 5 {
-		return hash[len(hash)-5:]
-	}
-
-	return libPath
+func getUID(lib pathIdentifier) string {
+	return lib.Key()[:5]
```
Comment (guyarb) [inline pkg/network/protocols/http/ebpf_ssl.go:432]: why 5?

File: pkg/network/protocols/http/shared_libraries.go [hunk 17]
```diff
@@ -35,118 +37,213 @@ func (l *libPath) Bytes() []byte {
...
+	}
+
 	go func() {
-		ticker := time.NewTicker(soSyncInterval)
-		defer ticker.Stop()
-		tickerProcess := time.NewTicker(soCheckProcessAliveInterval)
-		defer tickerProcess.Stop()
-		thisPID, _ := util.GetRootNSPID()
+		defer cleanupExit()
+		//TODO(nplanel) deregister uprobes() (unlikely error path)
```
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries.go:213]: what is that todo? is it still relevant? do you plan to add it as part of the current PR?

File: pkg/network/protocols/http/shared_libraries.go [hunk 18]
```diff
@@ -35,118 +37,213 @@ func (l *libPath) Bytes() []byte {
...
+	err := w.processMonitor.Initialize()
+	if err != nil {
+		log.Errorf("can't initialize process monitor %s", err)
+		return
+	}
+	cleanupExit, err := w.processMonitor.Subscribe(&monitor.ProcessCallback{
+		Event:    monitor.EXIT,
+		Metadata: monitor.ANY,
+		Callback: w.processExit,
+	})
```
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries.go:205]: why do you register only exit callback?

File: pkg/network/protocols/http/shared_libraries.go [hunk 19]
```diff
@@ -35,118 +37,213 @@ func (l *libPath) Bytes() []byte {
...
-				if int(lib.Pid) == thisPID {
-					event.Done()
-					break
+				var (
+					path    = lib.Bytes()
+					libPath = string(path)
+					procPid = fmt.Sprintf("%s/%d", w.procRoot, lib.Pid)
+					root    = procPid + "/root"
+					cwd     = procPid + "/cwd"
+				)
```
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries.go:230]:
```suggestion
					path    := lib.Bytes()
					libPath := string(path)
					procPid := fmt.Sprintf("%s/%d", w.procRoot, lib.Pid)
					root    := procPid + "/root"
					cwd     := procPid + "/cwd"
```

File: pkg/network/protocols/http/shared_libraries.go [hunk 20]
```diff
@@ -35,118 +37,213 @@ func (l *libPath) Bytes() []byte {
...
+				if strings.HasPrefix(libPath, "/proc/") { //lib.Pid == uint32(thisPID) { // don't scan ourself when we resolve offsets
+					continue
```
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries.go:232]: why do you skip?
the logic is unclear

File: pkg/network/protocols/http/shared_libraries.go [hunk 21]
```diff
@@ -35,118 +37,213 @@ func (l *libPath) Bytes() []byte {
...
-					var (
-						libPath = string(path)
-						pidPath = fmt.Sprintf("%s/%d", w.procRoot, lib.Pid)
-					)
-
-					// resolving paths is expensive so we cache the libraries we've already seen
-					k := seenKey{pidPath, libPath}
-					if _, ok := seen[k]; ok {
-						break
+					// use cwd of the process as root if the path is absolute
```
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries.go:240]:
```suggestion
					// use cwd of the process as root if the path is relative
```

File: pkg/network/protocols/http/shared_libraries.go [hunk 22]
```diff
@@ -160,221 +257,58 @@ func (w *soWatcher) Start() {
...
+	if err != nil {
+		// short living process can't hit here
+		// as we receive the openat() syscall info after receiving the EXIT netlink process
+		log.Debugf("can't create path identifier %s", err)
 		return
 	}
-
-	inode, err := getInode(libPath)
-	if err != nil {
+	if _, found := r.blocklistByID[pathID]; found {
```
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries.go:284]: should be after locking the mutex

File: pkg/network/protocols/http/shared_libraries.go [hunk 23]
```diff
@@ -160,221 +257,58 @@ func (w *soWatcher) Start() {
...
-	registration := newRegistration(inode, rule.unregisterCB)
-	registration.pids[pidPath] = struct{}{}
-	r.byPath[libPath] = registration
-	r.byInode[inode] = registration
-}
-
-func (r *soRegistry) unregister(libPath string) {
-	registration, ok := r.byPath[libPath]
-	if !ok {
+		r.blocklistByID[pathID] = struct{}{}
```
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries.go:305]: that's is an always increasing memory

File: pkg/network/protocols/http/shared_libraries.go [hunk 24]
```diff
@@ -160,221 +257,58 @@ func (w *soWatcher) Start() {
...
-		log.Debugf("error registering library=%s: %s", libPath, err)
-		if err := rule.unregisterCB(libPath); err != nil {
-			log.Debugf("unregisterCB %s : %s", libPath, err)
+	if err := rule.registerCB(pathID, root, libPath); err != nil {
+		log.Debugf("error registering library %s path %s by pid %d : %s", pathID.String(), hostLibPath, pid, err)
+		if rule.unregisterCB != nil {
+			if err := rule.unregisterCB(pathID); err != nil {
+				log.Debugf("unregisterCB library %s path %s : %s", pathID.String(), hostLibPath, err)
+			}
 		}
```
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries.go:302]: why do we need to unregister a callback that failed to register?

**Review by guyarb — CHANGES_REQUESTED** [review]

File: pkg/network/protocols/http/shared_libraries_test.go (hunk 13 above)
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries_test.go:87]: It is sleep, busybox change his personality based on the `basename(args[0])` this really useful in initrd or embedded environment (see busybox --install creating symlinks)
I remove the /sleep binary as it's not used directly, but sleep will be called from builtin ash command

File: pkg/network/protocols/http/shared_libraries.go [hunk 25]
```diff
@@ -35,118 +37,213 @@ func (l *libPath) Bytes() []byte {
...
+	binary.LittleEndian.PutUint64(b, p.dev)
+	binary.LittleEndian.PutUint64(b[8:], p.inode)
+	m := murmur3.Sum64(b)
+	bsum := make([]byte, 8)
+	binary.LittleEndian.PutUint64(bsum, m)
+	return base64.StdEncoding.EncodeToString(bsum)
+}
+
+// path must be an absolute path
+func newPathIdentifier(path string) (pi pathIdentifier, err error) {
```
[the remark this answers is not in the document]
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:61]: I'm using this form as I'm not creating a pointer of pathIdentifier as pathIdentifier object is comparable and is used mostly as a map key

File: pkg/network/protocols/http/shared_libraries.go (hunk 22 above)
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:284]: thanks

File: pkg/network/protocols/http/shared_libraries.go (hunk 23 above)
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:305]: yes but only on problematic path, so this scope is very limited
The only way to cleanup this entry would be the mTime check of the pathID, see jira USMO-172

File: pkg/network/protocols/http/shared_libraries.go (hunk 24 above)
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:302]: the old implementation use this mechanism of cleanup callback
As the register callback can failed to register any uprobe function, and they are registered one at a time in the kernel
Calling the unregister on error would cleanup already registered functions

Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:302]: I added a comment

File: pkg/network/protocols/http/ebpf_ssl.go (hunk 14 above)
Comment (nplanel) [reply pkg/network/protocols/http/ebpf_ssl.go:388]: we control root and path, and the kernel will solve the path himself when the file is used
filepath.Join() is nice but why spending cpu cleanup-ing a special path ?

File: pkg/network/protocols/http/shared_libraries.go (hunk 15 above)
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:57]: done

File: pkg/network/protocols/http/shared_libraries.go [hunk 26]
```diff
@@ -35,118 +37,213 @@ func (l *libPath) Bytes() []byte {
...
 
+type soRegistration struct {
+	pathID pathIdentifier
+	//mTime  uint64 TODO(nplanel) reload uprobes if the file changed (dnf upgrade)
+
+	refcount     int
+	unregisterCB func(pathIdentifier) error
+}
+
+func (r *soRegistration) Unregister() (cleanup bool) {
```
[the remark this answers is not in the document]
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:133]: done added comment

File: pkg/network/protocols/http/shared_libraries.go [hunk 27]
```diff
@@ -35,118 +37,213 @@ func (l *libPath) Bytes() []byte {
...
+	r.refcount--
+	if r.refcount > 0 {
+		return false
+	}
+	if r.unregisterCB != nil {
```
[the remark this answers is not in the document]
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:138]: the unit test is designed like that (soRule without unregister callback)

File: pkg/network/protocols/http/shared_libraries.go [hunk 28]
```diff
@@ -35,118 +37,213 @@ func (l *libPath) Bytes() []byte {
...
+		if err := r.unregisterCB(r.pathID); err != nil {
+			log.Debugf("unregisterCB %s : %s", r.pathID.String(), err)
```
[the remark this answers is not in the document]
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:140]: done

File: pkg/network/protocols/http/shared_libraries.go (hunk 7 above)
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:160]: there are no much to do there, I added log.Warn

File: pkg/network/protocols/http/shared_libraries.go (hunk 8 above)
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:187]: in case of multipe library version, but it's a corner case, I moved added a break

File: pkg/network/protocols/http/shared_libraries.go (hunk 18 above)
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:203]: we just need the EXIT process event to unregister the uprobes
uprobes are registered if the lib.so is openat() from the process
I did some test and scanning /proc/pid/smaps at Exec event, the event happen too early (the loader didn't got time to solve and load the .so)

File: pkg/network/protocols/http/shared_libraries.go (hunk 17 above)
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:211]: add cleanup all uprobes, calling all unregisterCB

File: pkg/network/protocols/http/shared_libraries.go (hunk 19 above)
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:228]: done

File: pkg/network/protocols/http/shared_libraries_test.go (hunk 13 above)
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries_test.go:89]: update: sleep need to be present in the path

**Review by xornivore — COMMENTED** [review]
:wave:  @nplanel I've removed the Processes team from reviewers since there doesn't seem to be shared code in this PR, but let me know if there's something you'd like me to take a closer look at 🙇

File: pkg/network/protocols/http/ebpf_ssl.go [hunk 29]
```diff
@@ -348,11 +354,42 @@ func (o *sslProgram) Stop() {
...
+		symbolMap, err := bininspect.GetAllSymbolsByName(elfFile, symbolsSet)
+		if err != nil {
+			return err
+		}
+		/* Best effort to resolve symbols, so we don't care about the error */
+		symbolMapBestEffort, _ := bininspect.GetAllSymbolsByName(elfFile, symbolsSetBestEffort)
+
+		for _, singleProbe := range probes {
+			_, isBestEffort := singleProbe.(*manager.BestEffort)
 			for _, selector := range singleProbe.GetProbesIdentificationPairList() {
```
Comment (p-lambert) [inline pkg/network/protocols/http/ebpf_ssl.go:393]: can we extract this to a helper function, add unit tests and document it? there is a lot of going on here and the code is a bit hard to follow :sweat:

File: pkg/network/protocols/http/ebpf_ssl.go [hunk 30]
```diff
@@ -291,12 +293,16 @@ func (o *sslProgram) ConfigureManager(m *errtelemetry.Manager) {
...
 	options.MapSpecEditors[sslSockByCtxMap] = manager.MapSpecEditor{
 		Type:       ebpf.Hash,
 		MaxEntries: uint32(o.cfg.MaxTrackedConnections),
 		EditorFlag: manager.EditMaxEntries,
 	}
+	options.MapSpecEditors[dispatcherConnectionProtocol] = manager.MapSpecEditor{
+		Type:       ebpf.Hash,
+		MaxEntries: uint32(o.cfg.MaxTrackedConnections),
+		EditorFlag: manager.EditMaxEntries,
+	}
```
Comment (p-lambert) [inline pkg/network/protocols/http/ebpf_ssl.go:305]: probably worth extracting this to a separate PR if this is the fix to the spinlock contention you mentioned

Comment (nplanel) [reply pkg/network/protocols/http/ebpf_ssl.go:305]: https://github.com/DataDog/datadog-agent/pull/15166

File: pkg/network/protocols/http/ebpf_ssl.go (hunk 29 above)
Comment (nplanel) [reply pkg/network/protocols/http/ebpf_ssl.go:393]: https://datadoghq.atlassian.net/browse/USMO-202

**Comment by nplanel** [comment]
@xornivore I think you left a LGTM previously, this initial PR was split in multiple PRs ... github ask you to re-approve even if he can't find the previous comment anymore

File: pkg/network/protocols/http/shared_libraries.go (hunk 18 above)
Comment (guyarb) [reply pkg/network/protocols/http/shared_libraries.go:205]: > I did some test and scanning /proc/pid/smaps at Exec event, the event happen too early (the loader didn't got time to solve and load the .so)

did we fix the issue?

File: pkg/network/protocols/http/shared_libraries.go [hunk 31]
```diff
@@ -160,221 +260,73 @@ func (w *soWatcher) Start() {
...
 	}
-
-	return updated
-}
-
-func (w *soWatcher) canonicalizePath(path string) string {
-	if w.hostMount != "" {
-		path = filepath.Join(w.hostMount, path)
+	if reg.Unregister() == true {
+		// we need to cleanup our entries as there are no more process using this ELF
```
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries.go:283]:
```suggestion
		// we need to cleanup our entries as there are no more processes using this ELF
```

**Review by guyarb — CHANGES_REQUESTED** [review]

File: pkg/network/protocols/http/ebpf_ssl.go [hunk 32]
```diff
@@ -416,14 +469,11 @@ func removeHooks(m *errtelemetry.Manager, probes []manager.ProbesSelector) func(
...
-	sum := murmur3.StringSum64(libPath)
-	hash := strconv.FormatInt(int64(sum), 16)
-	if len(hash) >= 5 {
-		return hash[len(hash)-5:]
-	}
-
-	return libPath
+// getUID() return a key of length 5 as the kernel uprobe registration path is limited to a length of 64
+// ebpf-manager/utils.go:GenerateEventName() MaxEventNameLen = 64
+// MAX_EVENT_NAME_LEN (linux/kernel/trace/trace.h)
```
[the remark this answers is not in the document]
Comment (nplanel) [reply pkg/network/protocols/http/ebpf_ssl.go:474]: it's a kind of space left here, as the full string of the eventName format 
`fmt.Sprintf("%s_%.*s_%s_%s", probeType, maxFuncNameLen, funcName, UID, attachPIDstr)`

[the remark this answers is not in the document]
Comment (nplanel) [reply pkg/network/protocols/http/ebpf_ssl.go:474]: and functionName is variable but with a minimum guarantee of 10 chars

File: pkg/network/protocols/http/shared_libraries.go [hunk 33]
```diff
@@ -35,118 +37,216 @@ func (l *libPath) Bytes() []byte {
...
+type pathIdentifier struct {
+	dev   uint64
+	inode uint64
+}
+
+func (p *pathIdentifier) String() string {
+	return fmt.Sprintf("dev/inode %d.%d/%d", unix.Major(p.dev), unix.Minor(p.dev), p.inode)
+}
+
+func (p *pathIdentifier) Key() string {
```
[the remark this answers is not in the document]
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:50]: done

File: pkg/network/protocols/http/shared_libraries.go (hunk 18 above)
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:205]: We are hooking only on the openat("libssl.so") event so we don't have the problem of early Exec

Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:205]: specifically for the Exec event, it's at the user callback to wait if needed, but the callback will be called (guarantee) with a /proc/pid/ existing if a metadata is required (read metadata.Name)

File: pkg/network/protocols/http/shared_libraries.go (hunk 31 above)
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:283]: done

File: pkg/network/tracer/tracer_usm_linux_test.go [hunk 34]
```diff
@@ -231,8 +231,10 @@ func testHTTPSLibrary(t *testing.T, fetchCmd []string) {
 	require.NoError(t, err)
 
 	// Run fetchCmd once to make sure the OpenSSL is detected and uprobes are attached
-	exec.Command(fetchCmd[0]).Run()
-	time.Sleep(2 * time.Second)
+	go func() {
+		pipesleep := fetchCmd[0] + " | sleep 2"
+		exec.Command("bash", "-c", pipesleep).Run()
+	}()
```
[the remark this answers is not in the document]
Comment (nplanel) [reply pkg/network/tracer/tracer_usm_linux_test.go]: I don't relay on this trick anymore
I simulate an open() of the libs to prehook

**Comment by xornivore** [comment]
> @xornivore I think you left a LGTM previously, this initial PR was split in multiple PRs ... github ask you to re-approve even if he can't find the previous comment anymore
I can remove myself since it appears you no longer need Processes' approval?

File: pkg/network/protocols/http/ebpf_ssl.go (hunk 32 above)
[the remark this answers is not in the document]
Comment (guyarb) [reply pkg/network/protocols/http/ebpf_ssl.go:474]: so that's what missing from the documentation, please add it

File: pkg/network/protocols/http/shared_libraries.go [hunk 35]
```diff
@@ -35,118 +37,224 @@ func (l *libPath) Bytes() []byte {
...
+type pathIdentifier struct {
+	dev   uint64
+	inode uint64
+}
+
+func (p *pathIdentifier) String() string {
+	return fmt.Sprintf("dev/inode %d.%d/%d", unix.Major(p.dev), unix.Minor(p.dev), p.inode)
+}
+
+// Key is unique (system wide) TLDR Base64(murmur3.Sum64(device, inode))
```
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries.go:50]:
```suggestion
// Key is a unique (system wide) TLDR Base64(murmur3.Sum64(device, inode))
```

File: pkg/network/protocols/http/shared_libraries.go [hunk 36]
```diff
@@ -35,118 +37,224 @@ func (l *libPath) Bytes() []byte {
...
+// It's construct based the device (minor, major) and inode of a file
+// murmur is non-crypto hashing
```
Comment (guyarb) [inline pkg/network/protocols/http/shared_libraries.go:52]:
```suggestion
// It composes the device (minor, major) and inode of a file
// murmur is a non-crypto hashing
```

Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:52]: done

File: pkg/network/protocols/http/shared_libraries.go (hunk 35 above)
Comment (nplanel) [reply pkg/network/protocols/http/shared_libraries.go:50]: done

File: pkg/network/tracer/tracer_usm_linux_test.go [hunk 37]
```diff
@@ -199,27 +199,30 @@ func TestHTTPSViaLibraryIntegration(t *testing.T) {
 					}
 					linked, _ := exec.Command(ldd, fetch).Output()
 
+					var prefechLibs []string
```
[the remark this answers is not in the document]
Comment (nplanel) [reply pkg/network/tracer/tracer_usm_linux_test.go:202]: done

**Review by guyarb — APPROVED** [review]
